Added sorted list implementation.

energyc0 opened the pull request:
Interactive sorted list exercise is done.

WoWaster commented on 06.10.2025/sortedList.h:
```diff
@@ -0,0 +1,53 @@
+
+typedef struct SortedList {
+    struct ListNode* _head;
+} SortedList;
```
Здесь аналогичное с соседним PR замечание: сделайте полноценный АТД и нейминг тоже слегка странный для Си.

WoWaster commented on 06.10.2025/sortedList.h:
```diff
@@ -0,0 +1,53 @@
+
+/*
+ * Free SortedList internals.
+ * Do not frees SortedList pointer.
```
```suggestion
 * Does not free SortedList pointer.
```
(иногда я обращаюсь в училку английского :) )

WoWaster commented on 06.10.2025/sortedList_exercise.c:
```diff
@@ -0,0 +1,135 @@
+static char skipSpaces(char* buf)
+{
+    for (; isspace(*buf); buf++)
+        ;
```
```suggestion
    while (isspace(*buf))
        buf++;
```

Как Вы понимаете, я от Вас с этим не отстану.

WoWaster commented on 06.10.2025/sortedList_exercise.c:
```diff
@@ -0,0 +1,135 @@
+static bool getNum(int* num)
+{
+    printf("Enter a number or 'q' to quit:\n");
+    char* data;
```
```suggestion
    char* data = NULL;
```

WoWaster commented on 06.10.2025/sortedList_exercise.c:
```diff
@@ -0,0 +1,135 @@
+ */
+static void addValue(SortedList* list)
+{
+    int num;
```
```suggestion
    int num = 0;
```

WoWaster commented on 06.10.2025/sortedList_exercise.c:
```diff
@@ -0,0 +1,135 @@
+    }
+    printf("Indexes start from 0.\n");
+
+    int idx;
```
```suggestion
    int idx = 0;
```

WoWaster commented on 06.10.2025/sortedList_exercise.c:
```diff
@@ -0,0 +1,135 @@
+        printf("Enter a command.\n");
+        if ((data = getInput()) == NULL
+            || !interactiveSortedList(&list, skipSpaces(data)))
+            break;
```
```suggestion
        data = getInput();
        if (data == NULL)
            break;
        bool res = interactiveSortedList(&list, skipSpaces(data));
        if (!res)
            break;
```

WoWaster commented on 06.10.2025/sortedList_exercise.c:
```diff
@@ -0,0 +1,135 @@
+        char* data;
+        printf("Enter a command.\n");
+        if ((data = getInput()) == NULL
+            || !interactiveSortedList(&list, skipSpaces(data)))
+            break;
+    }
+    return 0;
```
А тут Вы забыли почистить память после использования списка :(

WoWaster commented on 06.10.2025/sortedList.c:
```diff
@@ -0,0 +1,145 @@
+
+static ListNode* allocListNode(int value);
+
+ListNode* allocListNode(int value)
```
А такое дублирование правда нужно?

WoWaster commented on 06.10.2025/sortedList.c:
```diff
@@ -0,0 +1,145 @@
+    if (pList->_head == NULL) {
+        pList->_head = newNode;
+        return true;
+    } else if (pList->_head->value >= value) {
```
```suggestion
    }
    
    if (pList->_head->value >= value) {
```

Так было бы понятнее, что мы сначала разибраем крайние случаи, а потом вставляем "на общих основаниях"

WoWaster commented on 06.10.2025/sortedList.c:
```diff
@@ -0,0 +1,145 @@
+
+    ListNode* ptr = pList->_head;
+    for (; ptr->next && ptr->next->value < value; ptr = ptr->next)
+        ;
```
```suggestion
    while (ptr->next != NULL && ptr->next->value < value)
        ptr = ptr->next;
```

Можно я поверю, что Вы меня поняли и не буду так же комментировать каждый следующий `for`?

energyc0 committed:
Refactored code.
Explicit initialization, sorted list is now an abstract data type, changed 'for' loops to 'while'.

energyc0 committed:
Added sorted list implementation.
Interactive sorted list exercise is done.

energyc0 commented:
Rebased main for resolving conflicts.
